Table: Add colors block support

aaronrobertshaw opened the pull request:
## Description
Replace ad-hoc background color selection for tables with colors block support.

- Allows selection of text and background colors
- With recent feature to skip serialization of block support classes/styles, this can now target the table element as required
- Previous color selections are migrated via a deprecation to the appropriate style attribute
- Includes updated fixture tests 

## How has this been tested?
Manually and via fixtures tests using `npm run test-unit` 

#### Test Instructions
1. Create a new post, add a table block and select a background color for the table
2. Save the post
3. Checkout this PR branch, rebuild, then reload the editor
4. Confirm the table block migrated successfully and the previous color selection is still present
5. Select the table block and modify the background and font colors
6. Save post and confirm selections are rendered correctly on the frontend

## Screenshots <!-- if applicable -->
<img src="https://user-images.githubusercontent.com/60436221/114511998-f299ac80-9c7b-11eb-9da3-7346c74b0c51.gif" width="400"/>

## Types of changes
New feature.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [x] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

aaronrobertshaw committed:
Add color block support to table block

aaronrobertshaw committed:
Update fixtures to match table block deprecations

aaronrobertshaw commented:
@nosolosw Do you have any thoughts on standardising retrieval of block support provided classes and styles when we are leveraging the "skip serialisation" feature?

With the ability to pick and choose which inner elements we apply these block support props to, it seems like something we'll be doing more of in the near future. In this PR's case, I've leveraged the util from the button block that was extracted from the colors block support hook.

oandregal commented:
> @nosolosw Do you have any thoughts on standardising retrieval of block support provided classes and styles when we are leveraging the "skip serialisation" feature?

Yeah, we need that extracted to a utility. It seems this util can be something like `__experimentalGetClassesAndStyleForColors`. Somewhere in the `block-editor/hooks` should be a good place. I'd think this is best done in a separate PR before this one.

(Personally, as a reviewer, I welcome many small PRs rather than a big one. Small slots for reviews work best for me, as it allows me to help others while also advance my work without having to block half a day for a single review).

oandregal commented on packages/block-library/src/table/block.json:
```diff
@@ -125,6 +122,11 @@
+		"color": {
+			"__experimentalSkipSerialization": true,
+			"gradients": true,
+			"link": false
```
Probably not necessary at the moment as the existing `getColorAndStyleProps` doesn't process link color, but it brings the question: shouldn't it?

aaronrobertshaw replied:
I believe the intent of the current `getColorAndStyleProps` was to only bring in what was needed for the button block. That happened to line up with what was needed here for the table block. 

More than likely, a generic utility would need to take into account the current block's configuration of block supports.

talldan commented:
When testing this, will need to make sure the 'stripes' block style is well tested and works as expected. Also worth testing with table headers and footers active.

I did a quick test, and looks like text color doesn't seem to work for 'stripes'.

aaronrobertshaw commented:
> Yeah, we need that extracted to a utility. It seems this util can be something like __experimentalGetClassesAndStyleForColors. Somewhere in the block-editor/hooks should be a good place. I'd think this is best done in a separate PR before this one.

My thinking was as more block support features are implemented and there are more cases of skipping serialization. Having a utility to retrieve the css classes and inline styles across block support features would be good if possible.

`__experimentalGetBlockSupportClassesAndStyles( [ 'colors', 'border' ] )`

That might be over-the-top though in hindsight.

> (Personally, as a reviewer, I welcome many small PRs rather than a big one. Small slots for reviews work best for me, as it allows me to help others while also advance my work without having to block half a day for a single review).

I'm happy to create a couple of PRs, extracting `getColorAndStyleProps` into a utility in `block-editor/hooks`, one to update the button block accordingly, then updating this as well.

aaronrobertshaw commented:
> I did a quick test, and looks like text color doesn't seem to work for 'stripes'.

@talldan Any chance you can let me know which theme you were using when you tested? 

The "stripes" block style works ok for me with the custom colors. When you found it not to work, was that related to the arbitrary background color for the striped rows clashing with a selected text color?

![TableColors-2](https://user-images.githubusercontent.com/60436221/114672321-5db0b500-9d48-11eb-8741-d69019781c66.gif)

talldan commented:
@aaronrobertshaw I did some further testing, and I'm seeing that text colors don't work when a background color is selected with the stripes style:
![table-colors](https://user-images.githubusercontent.com/677833/114803182-99439180-9dd1-11eb-89b9-f0291d6719bc.gif)

Using Twenty Twenty One. I think it's a theme issue, it works ok with editor theme styles disabled in the settings. Is that something you can see as well? If so, it'd be good to follow up with a trac ticket.

aaronrobertshaw commented:
> Using Twenty Twenty One. I think it's a theme issue, it works ok with editor theme styles disabled in the settings. Is that something you can see as well? If so, it'd be good to follow up with a trac ticket.

From memory, I was likely using the TT1 Blocks theme. At least thats the first explanation that comes to mind as the GIF in my previous comment shows the font colors being applied correctly both in the editor and frontend. 

The standard Twenty Twenty One theme may not be qualifying for the `has-<color>-text-color` classes to be generated. The `getColorAndStyleProps` function currently forces inline styles for the colors in the editor.

I'll test further and see if I can confirm that.

aaronrobertshaw commented:
After further testing, I found two themes that introduced conflicting styles preventing the selected table colors from displaying correctly. These themes were:

- TwentyTwentyOne
- Maywood

![TableColors-ThemeIssue](https://user-images.githubusercontent.com/60436221/114814228-1a614f80-9df7-11eb-9277-2d945c56ca16.gif)

In other themes, the text color was displayed correctly in both the editor and the frontend. These themes included:

- TT1 Blocks (block based TwentyTwentyOne)
- TwentyTwenty
- TwentyTwenty (block based)
- Q
- Varia
- Seedlet
- Hever

> If so, it'd be good to follow up with a trac ticket.

If the approach to applying block support styles in this PR is accepted I can create any trac tickets needed so themes get updated to honour users' color selections.

talldan reviewed (approved):
Thanks for investigating those issues. Would be great to follow up with trac tickets.

The changes here look good, and users will be very happy to see this shipped. Nice work 🎉

talldan commented:
To get this merged, looks like it may require a rebase and reformat of the json files as per https://github.com/WordPress/gutenberg/pull/30714.

aaronrobertshaw commented:
> The changes here look good, and users will be very happy to see this shipped. Nice work 🎉
> To get this merged, looks like it may require a rebase and reformat of the json files as per #30714.

Thanks for all the effort on the reviews 👍 

As part of addressing the earlier suggestion to move the `getColorAndStyleProps` function into a new utility available via the block support hooks, I've rebased and updated this PR to utilise that. https://github.com/WordPress/gutenberg/pull/30869 

My apologies for the updates post approval!

talldan commented:
> My apologies for the updates post approval!

No worries! I think you could've merged and then followed up with the other change in another PR, but it's ok to do it this way as well.

I'm not all that familiar with the support hook, so would appreciate if @nosolosw could give the thumbs up to that change.

aaronrobertshaw commented:
Thanks for taking another look at this one 👍 

> I'm not all that familiar with the support hook, so would appreciate if @nosolosw could give the thumbs up to that change.

The [color block support utilities](https://github.com/WordPress/gutenberg/pull/30869) have now been merged as well as [updates to the button block](https://github.com/WordPress/gutenberg/pull/30870) to use them. Hopefully, this should only need a quick rubber stamp from here.

jorgefilipecosta commented on packages/block-library/src/table/edit.js:
```diff
@@ -58,29 +57,6 @@ import {
 	isEmptyTableSection,
 } from './state';
 
-const BACKGROUND_COLORS = [
```
There was a requirement that the table block given its specifies had a specific color palette different from the other blocks, instead of using the default or theme colors. This requirement added some complexity and was the reason createCustomColorsHOC was implemented.
This PR removes that mechanism and makes the table block use the same colors as the other blocks. Is this change intentional and the requirement of having specific colors was removed?


I guess we could easily keep the specific colors by adding a table block specific color palette to lib/experimental-default-theme.json. Would that make sense?

aaronrobertshaw replied:
The idea of having a custom color palette for the table only might work. I wonder if it is missing an opportunity though.

The table block was identified as benefiting from the colors block support in https://github.com/WordPress/gutenberg/issues/28913. The colors block support would provide more control for users including the ability to set text color. My understanding was the block support also then includes contrast checking for the background and text colors.

There is also [this issue](https://github.com/WordPress/gutenberg/issues/19659) that discusses the need for additional table color options.

jorgefilipecosta commented on packages/block-library/src/table/deprecated.js:
```diff
@@ -6,13 +6,261 @@ import classnames from 'classnames';
+// As the previous arbitrary colors won't match theme color palettes, the hex
+// value will be mapped to the style.color.background attribute as if it was
+// a custom color selection.
+const oldColors = {
```
We are making every table block that exists use inline styles instead of a color class. But we will still need to keep the old classes so the blocks properly work on the frontend even if they are never opened on the editor.

Could we leave the attributes as they were instead (keeping the classes and named attributes)? I guess we could keep the old colors as the default table block palette on lib/experimental-default-theme.json. Given that the classes still exist and will always be preset I guess the blocks should still look ok without any migration.

aaronrobertshaw replied:
> We are making every table block that exists use inline styles instead of a color class. But we will still need to keep the old classes so the blocks properly work on the frontend even if they are never opened on the editor.

If the block's are not opened in the editor and therefore not migrated. They would still keep the old CSS classes. This is why those classes still exist within the `style.css` file. 

> Could we leave the attributes as they were instead (keeping the classes and named attributes)? I guess we could keep the old colors as the default table block palette on lib/experimental-default-theme.json. Given that the classes still exist and will always be preset I guess the blocks should still look ok without any migration.

If the table block is given its own custom color palette, as you suggested, with the old colors under the same names, their CSS classes should be unchanged. With that approach, the original `backgroundColor` attribute could remain as is, instead of migrating it to the `style.color.background` attribute.

jorgefilipecosta commented on packages/e2e-tests/fixtures/blocks/core__table__deprecated-1.html:
```diff
@@ -1,3 +1,3 @@
-<!-- wp:core/table -->
-<table class="wp-block-table"><thead><tr><th>Version</th><th>Musician</th><th>Date</th></tr></thead><tbody><tr><td><a href="https://wordpress.org/news/2003/05/wordpress-now-available/">.70</a></td><td>No musician chosen.</td><td>May 27, 2003</td></tr><tr><td><a href="https://wordpress.org/news/2004/01/wordpress-10/">1.0</a></td><td>Miles Davis</td><td>January 3, 2004</td></tr><tr><td>Lots of vers ...
-<!-- /wp:core/table -->
+<!-- wp:table {"backgroundColor":"subtle-pale-blue"} -->
```
What makes us need to change and existing deprecation? Could we leave this deprecation without changes and create a new deprecation specific to the new case we want to test?

aaronrobertshaw replied:
My understanding is that the ideal order for deprecations is reverse chronological. In keeping with past advice to that effect, I added the deprecation for the colors block support changes to the beginning of the deprecated array.

I regenerated the deprecation fixtures so they still aligned. That is, the first deprecation in the array matches the fixture named `*_deprecated-1.html`.

Deprecations in general feel very counter intuitive. So making the fixtures not align didn't see like an option. Leaving a choice to either:
- Append the new deprecation to the array and only add new fixtures (which went against past recommendations for deprecation order)
- Prepend the new deprecation and regenerate fixtures (the downside being inconvenience when reviewing).

I went with the later as it was a one time cost and got the slight optimisation of the reverse-chronologically order deprecations.

jorgefilipecosta commented on packages/block-library/src/table/edit.js:
```diff
@@ -499,27 +473,14 @@ function TableEdit( {
-								value: backgroundColor.color,
-								onChange: setBackgroundColor,
-								label: __( 'Background color' ),
-								disableCustomColors: true,
```
Previously custom colors were intentionally disabled for the table block. This PR enables custom colors for the table block. Is this change intentional? Should we keep them disabled using lib/experimental-default-theme.json?

aaronrobertshaw replied:
With the adoption of the colors block support increasing the user's control over colors including font color, it seemed like this was the preferred direction to go. More so when taking into account some of the conversation in https://github.com/WordPress/gutenberg/issues/19659.

aaronrobertshaw commented:
Thank you for the great feedback @jorgefilipecosta 👍 

It would be great to get some input from others regarding whether we should maintain a custom color palette for the table block as suggested as well as allow custom colors. @talldan and @nosolosw do you have any ideas or preferences on this?

talldan commented:
I would defer to the conversation on #19659, where it has been made very clear that the current limited palette is a problem for users.

See @kjellr's comment here:
https://github.com/WordPress/gutenberg/issues/19659#issuecomment-794232681

I think this PR would be a good first step, but there are additional challenges that I think it would be good to devote some time to solving:
- Users can't specify different colors for table headers and footers, which is a pretty common thing to do.
- The striped style generally causes contrast issues, since the table has two background colors when this is active, but the user can only specify one of those background colors and one text color. Generally it's easiest to go for a lighter background color, but themes often don't provide a decent range of options.

oandregal commented on packages/block-editor/src/hooks/use-color-props.js:
```diff
@@ -0,0 +1,118 @@
+/**
```
Wasn't this merged as part of https://github.com/WordPress/gutenberg/pull/30869? I guess this needs a rebase.

aaronrobertshaw replied:
✅ Rebased.

oandregal commented on packages/block-library/src/table/block.json:
```diff
@@ -125,6 +122,10 @@
 	"supports": {
 		"anchor": true,
 		"align": true,
+		"color": {
```
I've tested it and the background color works as in `trunk`. That part is ready to land in my view. 

However, I'm not sure about text color for this block: it affects both text and the table borders. Shouldn't those things be configurable separately?

One thing we could do to land the parts of this PR that are ready is this: only enable background (and gradients) for now and prepare a follow-up to explore the relationship between text & border gathering more design feedback.

aaronrobertshaw replied:
> However, I'm not sure about text color for this block: it affects both text and the table borders. Shouldn't those things be configurable separately?

This comes from the active theme. Several themes set the border color to `initial` which means they get set to `currentcolor` and take on the font color. With the new border color block support it could be configurable separately. Which as you say, is better done in a follow up PR.

I thought the greater concern is ensuring sufficient contrast between the background and text colors. If we adopt the colors block support for the background but not text color, do we just create a bigger issue? Is that acceptable if we are looking to dial back this PR to only update background color?

oandregal replied:
Thought about this more: the current behavior is also text + border, so we should be fine.

I guess what I'm uncomfortable with is that this block has a larger design surface than many other blocks and it doesn't look super great at the moment (two background colors -with stripes-, border & text, etc). These are things that Dan brought up already https://github.com/WordPress/gutenberg/pull/30791#issuecomment-823815094 so let's merge and iterate.

aaronrobertshaw replied:
Sounds like a plan 👍 

I'll merge and take a look at some options to address https://github.com/WordPress/gutenberg/pull/30791#issuecomment-823815094.

oandregal commented:
In terms of how this uses the block supports: it's fine, although it looks like we need to rebase to get the latest utilities (and remove those changes from this PR).

I suggested at https://github.com/WordPress/gutenberg/pull/30791#discussion_r620218821 that we start only with background for this PR (which works as of today, just with more theme colors) and tackle text/border separately.

oandregal commented:
By the way, I've noticed the dropdown for the default style looks a bit odd (both Firefox and Chrome on Linux):

![Captura de ecrã de 2021-04-26 16-22-52](https://user-images.githubusercontent.com/583546/116098956-c898af00-a6ab-11eb-96dd-029db4abe805.png)

aaronrobertshaw commented:
> By the way, I've noticed the dropdown for the default style looks a bit odd (both Firefox and Chrome on Linux):

Unfortunately, I don't have a Linux VM handy at the moment to test. It appears OK for me across browsers under Mac and Windows using a few different themes.

Is it just this PR you noticed the issue with and is it still an issue after the rebase?

oandregal commented:
> Is it just this PR you noticed the issue with and is it still an issue after the rebase?

Oh, sorry, hadn't tested on `trunk`: I just did and it also happens over there.

oandregal reviewed (approved):
Tested this a bit and works as expected (tested deprecations with old colors and are properly migrated to the new system). From the point of view of the block supports, it's also ready to land.

aaronrobertshaw commented:
I've created https://github.com/WordPress/gutenberg/issues/31261 to track the desired follow ups mentioned in https://github.com/WordPress/gutenberg/pull/30791#issuecomment-823815094
